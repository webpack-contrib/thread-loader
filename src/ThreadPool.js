import * as path from 'path';
import { Worker } from 'worker_threads';
import { wrap, proxy } from 'comlink';
import { Subject, asapScheduler, zip } from 'rxjs';
import { filter, map, mergeMap, observeOn, takeUntil } from 'rxjs/operators';

const nodeEndpoint = require('comlink/dist/umd/node-adapter');

/**
 * Create worker object wraps native threads with proxy interfaces.
 */
const createWorker = () => {
  const worker = new Worker(path.resolve(__dirname, './workerEntryPoint.js')});
  const workerProxy = wrap(nodeEndpoint(worker));
  worker.unref();

  let disposed = false;

  return {
    disposed,
    workerProxy,
    close: () => {
      disposed = true;
      return new Promise((resolve) => {
        worker.once('exit', () => {
          workerProxy[releaseProxy]();
          resolve();
        });

        workerProxy.close();
      });
    },
  };
};

/**
 * Create comlink proxy-wrapped transferrable object from given
 * worker data context.
 *
 * Each loader's data context includes Webpack::loader::LoaderContext
 * have various functions. This marshall splits object between POJO to functions,
 * then wrap all functions into comlink proxy to avoid cloning attempt.
 *
 * In marshalled object, non-proxied (POJO) contains all keys of proxied fn property
 * as iterating & gettings keys to proxy object is bit tricky.
 *
 * Note `workerEntryPoint` have additional handling for some edge cases as well.
 */
const marshallWorkerDataContext = context =>
  Object.entries(context).reduce(
    (acc, [key, value]) => {
      if (typeof value === 'function') {
        acc[1][key] = proxy(value);
        acc[0].proxyFnKeys.push(key);
      } else {
        acc[0][key] = value;
      }

      return acc;
    },
    [{ proxyFnKeys: [] }, {}],
  );

export default class ThreadPool {
  constructor(maxWorkers) {
    this.maxWorkers = maxWorkers;
    this.taskCount = 0;
    this.timeoutId = null;

    this.taskQueue = new Subject();
    this.workerQueue = new Subject();
    this.disposeAwaiter = new Subject();

    this.workerPool = [...new Array(maxWorkers)].map(() => createWorker());

    this.poolSubscription = this.startPoolScheduler();
  }

  /**
   * Ask all threads to exit once queued task completes.
   */
  async closeWorkers() {
    let worker = this.workerPool.shift();
    while (worker) {
      if (!worker.disposed) {
        await worker.close();
      }
      worker = this.workerPool.shift();
    }
  }

  /**
   * Try to exit existing workers when there's no task scheduled within timeout (2sec)
   * If there is any running task when timeout reaches extend timeout to next timeout tick.
   */
  async scheduleTimeout() {
    if (this.timeoutId) {
      clearTimeout(this.timeoutId);
      this.timeoutId = null;
    }

    this.timeoutId = setTimeout(async () => {
      if (this.taskCount === 0) {
        await this.closeWorkers();
      } else {
        this.scheduleTimeout();
      }
    }, 2000);
  }

  /**
  * Run task via worker, raises timeoutError if worker does not respond in timeout period (10sec).
  * Most cases this happens when task is scheduled into disposed worker which released complink proxy already.
  */
  tryRunTaskWithTimeout(
    worker,
    id,
    context,
    proxyContext,
  ) {
    let runTaskTimeoutId = null;

    return new Promise((resolve, reject) => {
      runTaskTimeoutId = setTimeout(() => {
        if (worker.disposed) {
          this.workerPool.splice(this.workerPool.indexOf(worker), 1);
        }
        reject({ timeout: true });
      }, 10000);

      worker.workerProxy.run({ id }, context, proxyContext).then(
        (result) => {
          if (runTaskTimeoutId) {
            clearTimeout(runTaskTimeoutId);
            runTaskTimeoutId = null;
          }
          resolve(result);
        },
        (err) => {
          if (runTaskTimeoutId) {
            clearTimeout(runTaskTimeoutId);
            runTaskTimeoutId = null;
          }
          reject(err);
        },
      );
    });
  }

  /**
   * Actual task scheduler.
   */
  startPoolScheduler() {
    const sub = zip(
      // Each time new task is scheduled, reset timeout for close worker.
      // If this task is scheduled after timeout, it will reinstall worker threads.
      this.taskQueue.pipe(
        map((v) => {
          this.scheduleTimeout();

          if (this.workerPool.length < this.maxWorkers) {
            for (let idx = this.workerPool.length; idx < this.maxWorkers; idx++) {
              const worker = createWorker();
              this.workerPool.push(worker);
              this.workerQueue.next(worker);
            }
          }
          return v;
        }),
      ),
      this.workerQueue.pipe(
        filter(x => !x.disposed),
        map((v, i) => v),
      ),
    )
      .pipe(
        observeOn(asapScheduler),
        takeUntil(this.disposeAwaiter),
        mergeMap(async ([task, worker]) => {
          if (!worker || worker.disposed) {
            this.taskQueue.next(task);
          }
          const { context, proxyContext, id } = task;

          try {
            const result = await this.tryRunTaskWithTimeout(worker, id, context, proxyContext);
            task.onComplete(result);
            return { id, value: true };
          } catch (err) {
            if (!!err && err.timeout) {
              this.taskQueue.next(task);
            } else {
              task.onError(err);
            }

            return { id, value: false };
          }
        }),
      )
      .subscribe(
        (x) => {
          this.taskCount--;

          // Each time task completes, queue new worker to let zip operator picks up task / worker pair
          const worker = this.workerPool[(x.id + 1) % this.maxWorkers];
          this.workerQueue.next(worker);
        },
        null,
        () => {
          this.closeWorkers();
        },
      );

    // Queue all workers when starting scheduler
    this.workerPool.forEach(function initialQueue(w) { this.workerQueue.next(w); });

    return sub;
  }

  warmup() {
    /* noop */
  }

  run(data, callback) {
    if (this.timeoutId) {
      clearTimeout(this.timeoutId);
      this.timeoutId = null;
    }

    ++this.taskCount;

    const [normalContext, proxyContext] = marshallWorkerDataContext(context);
    this.taskQueue.next({
      context: normalContext,
      // Wrap whole object into proxy again, otherwise worker will try clone
      proxyContext: proxy(proxyContext),
      onComplete: callback,
      onError: callback,
    });
  }
}
