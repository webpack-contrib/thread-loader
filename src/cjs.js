function threadLoader() {
  throw new Error('This function should never be invoked');
}
Object.assign(threadLoader, require('./index'));

module.exports = threadLoader;
