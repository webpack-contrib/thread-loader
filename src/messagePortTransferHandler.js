import { expose, proxyMarker, transferHandlers, wrap } from 'comlink';
import { MessageChannel } from 'worker_threads';

const nodeEndpoint = require('comlink/dist/umd/node-adapter');

/**
 * Override comlink's default proxy handler to use Node endpoints
 * https://github.com/GoogleChromeLabs/comlink/issues/313
 */
const setupTransferHandler = () => {
  transferHandlers.set('proxy', {
    canHandle: obj => obj && obj[proxyMarker],
    serialize: (obj) => {
      const { port1, port2 } = new MessageChannel();
      expose(obj, nodeEndpoint(port1));
      return [port2, [port2]];
    },
    deserialize: (port) => {
      port = nodeEndpoint(port);
      port.start();
      return wrap(port);
    },
  });
};

export { setupTransferHandler };
