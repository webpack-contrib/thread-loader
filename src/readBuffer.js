export default function readBuffer(pipe, length, callback) {
  if (length === 0) {
    callback(null, Buffer.alloc(0));
    return;
  }

  let remainingLength = length;

  const buffers = [];

  const readChunk = () => {
    const onChunk = (arg) => {
      let chunk = arg;
      let overflow;

      if (chunk.length > remainingLength) {
        overflow = chunk.slice(remainingLength);
        chunk = chunk.slice(0, remainingLength);
        remainingLength = 0;
      } else {
        remainingLength -= chunk.length;
      }

      buffers.push(chunk);

      if (remainingLength === 0) {
        pipe.removeListener('data', onChunk);
        pipe.pause();

        if (overflow) {
          pipe.unshift(overflow);
        }

        callback(null, Buffer.concat(buffers, length));
      }
    };

    pipe.on('data', onChunk);
    pipe.resume();
  };
  readChunk();
}
