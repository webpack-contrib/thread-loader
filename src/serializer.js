export function replacer(_key, value) {
  if (value instanceof RegExp) {
    return {
      __serialized_type: 'RegExp',
      source: value.source,
      flags: value.flags,
    };
  }
  return value;
}

export function reviver(_key, value) {
  if (typeof value === 'object' && value !== null) {
    // eslint-disable-next-line no-underscore-dangle
    if (value.__serialized_type === 'RegExp') {
      return new RegExp(value.source, value.flags);
    }
  }

  return value;
}
