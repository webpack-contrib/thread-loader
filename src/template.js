// TODO export it from webpack

const { basename, extname } = require('path');
const util = require('util');

const { Chunk } = require('webpack');
const { Module } = require('webpack');
const { parseResource } = require('webpack/lib/util/identifier');

const REGEXP = /\[\\*([\w:]+)\\*\]/gi;

/**
 * @param {string | number} id id
 * @returns {string | number} result
 */
const prepareId = (id) => {
  if (typeof id !== 'string') return id;

  if (/^"\s\+*.*\+\s*"$/.test(id)) {
    const match = /^"\s\+*\s*(.*)\s*\+\s*"$/.exec(id);

    return `" + (${
      /** @type {string[]} */ (match)[1]
    } + "").replace(/(^[.-]|[^a-zA-Z0-9_-])+/g, "_") + "`;
  }

  return id.replace(/(^[.-]|[^a-zA-Z0-9_-])+/g, '_');
};

/**
 * @callback ReplacerFunction
 * @param {string} match
 * @param {string | undefined} arg
 * @param {string} input
 */

/**
 * @param {ReplacerFunction} replacer replacer
 * @param {((arg0: number) => string) | undefined} handler handler
 * @param {AssetInfo | undefined} assetInfo asset info
 * @param {string} hashName hash name
 * @returns {ReplacerFunction} hash replacer function
 */
const hashLength = (replacer, handler, assetInfo, hashName) => {
  /** @type {ReplacerFunction} */
  const fn = (match, arg, input) => {
    let result;
    const length = arg && Number.parseInt(arg, 10);

    if (length && handler) {
      result = handler(length);
    } else {
      const hash = replacer(match, arg, input);

      result = length ? hash.slice(0, length) : hash;
    }
    if (assetInfo) {
      // eslint-disable-next-line no-param-reassign
      assetInfo.immutable = true;
      if (Array.isArray(assetInfo[hashName])) {
        // eslint-disable-next-line no-param-reassign
        assetInfo[hashName] = [...assetInfo[hashName], result];
      } else if (assetInfo[hashName]) {
        // eslint-disable-next-line no-param-reassign
        assetInfo[hashName] = [assetInfo[hashName], result];
      } else {
        // eslint-disable-next-line no-param-reassign
        assetInfo[hashName] = result;
      }
    }
    return result;
  };

  return fn;
};

/** @typedef {(match: string, arg?: string, input?: string) => string} Replacer */

/**
 * @param {string | number | null | undefined | (() => string | number | null | undefined)} value value
 * @param {boolean=} allowEmpty allow empty
 * @returns {Replacer} replacer
 */
const replacer = (value, allowEmpty) => {
  /** @type {Replacer} */
  const fn = (match, arg, input) => {
    if (typeof value === 'function') {
      // eslint-disable-next-line no-param-reassign
      value = value();
    }
    // eslint-disable-next-line no-undefined
    if (value === null || value === undefined) {
      if (!allowEmpty) {
        throw new Error(
          `Path variable ${match} not implemented in this context: ${input}`,
        );
      }

      return '';
    }

    return `${value}`;
  };

  return fn;
};

const deprecationCache = new Map();
const deprecatedFunction = (() => () => {})();
/**
 * @param {Function} fn function
 * @param {string} message message
 * @param {string} code code
 * @returns {function(...any[]): void} function with deprecation output
 */
const deprecated = (fn, message, code) => {
  let d = deprecationCache.get(message);
  // eslint-disable-next-line no-undefined
  if (d === undefined) {
    d = util.deprecate(deprecatedFunction, message, code);
    deprecationCache.set(message, d);
  }
  return (...args) => {
    d();
    return fn(...args);
  };
};

/** @typedef {string | function(PathData, AssetInfo=): string} TemplatePath */

/**
 * @param {TemplatePath} path the raw path
 * @param {PathData} data context data
 * @param {AssetInfo | undefined} assetInfo extra info about the asset (will be written to)
 * @returns {string} the interpolated path
 */
const replacePathVariables = (path, data, assetInfo) => {
  const { chunkGraph } = data;

  /** @type {Map<string, Function>} */
  const replacements = new Map();

  // Filename context
  //
  // Placeholders
  //
  // for /some/path/file.js?query#fragment:
  // [file] - /some/path/file.js
  // [query] - ?query
  // [fragment] - #fragment
  // [base] - file.js
  // [path] - /some/path/
  // [name] - file
  // [ext] - .js
  if (typeof data.filename === 'string') {
    const { path: file, query, fragment } = parseResource(data.filename);

    const ext = extname(file);
    const base = basename(file);
    const name = base.slice(0, base.length - ext.length);
    // eslint-disable-next-line no-shadow
    const path = file.slice(0, file.length - base.length);

    replacements.set('file', replacer(file));
    replacements.set('query', replacer(query, true));
    replacements.set('fragment', replacer(fragment, true));
    replacements.set('path', replacer(path, true));
    replacements.set('base', replacer(base));
    replacements.set('name', replacer(name));
    replacements.set('ext', replacer(ext, true));
    // Legacy
    replacements.set(
      'filebase',
      deprecated(
        replacer(base),
        '[filebase] is now [base]',
        'DEP_WEBPACK_TEMPLATE_PATH_PLUGIN_REPLACE_PATH_VARIABLES_FILENAME',
      ),
    );
  }

  // Compilation context
  //
  // Placeholders
  //
  // [fullhash] - data.hash (3a4b5c6e7f)
  //
  // Legacy Placeholders
  //
  // [hash] - data.hash (3a4b5c6e7f)
  if (data.hash) {
    const hashReplacer = hashLength(
      replacer(data.hash),
      data.hashWithLength,
      assetInfo,
      'fullhash',
    );

    replacements.set('fullhash', hashReplacer);

    // Legacy
    replacements.set(
      'hash',
      deprecated(
        hashReplacer,
        '[hash] is now [fullhash] (also consider using [chunkhash] or [contenthash], see documentation for details)',
        'DEP_WEBPACK_TEMPLATE_PATH_PLUGIN_REPLACE_PATH_VARIABLES_HASH',
      ),
    );
  }

  // Chunk Context
  //
  // Placeholders
  //
  // [id] - chunk.id (0.js)
  // [name] - chunk.name (app.js)
  // [chunkhash] - chunk.hash (7823t4t4.js)
  // [contenthash] - chunk.contentHash[type] (3256u3zg.js)
  if (data.chunk) {
    const { chunk } = data;

    const { contentHashType } = data;

    const idReplacer = replacer(chunk.id);
    const nameReplacer = replacer(chunk.name || chunk.id);
    const chunkhashReplacer = hashLength(
      replacer(chunk instanceof Chunk ? chunk.renderedHash : chunk.hash),
      // eslint-disable-next-line no-undefined
      'hashWithLength' in chunk ? chunk.hashWithLength : undefined,
      assetInfo,
      'chunkhash',
    );
    const contenthashReplacer = hashLength(
      replacer(
        data.contentHash ||
          (contentHashType &&
            chunk.contentHash &&
            chunk.contentHash[contentHashType]),
      ),
      data.contentHashWithLength ||
        ('contentHashWithLength' in chunk && chunk.contentHashWithLength
          ? chunk.contentHashWithLength[/** @type {string} */ (contentHashType)]
          : // eslint-disable-next-line no-undefined
            undefined),
      assetInfo,
      'contenthash',
    );

    replacements.set('id', idReplacer);
    replacements.set('name', nameReplacer);
    replacements.set('chunkhash', chunkhashReplacer);
    replacements.set('contenthash', contenthashReplacer);
  }

  // Module Context
  //
  // Placeholders
  //
  // [id] - module.id (2.png)
  // [hash] - module.hash (6237543873.png)
  //
  // Legacy Placeholders
  //
  // [moduleid] - module.id (2.png)
  // [modulehash] - module.hash (6237543873.png)
  if (data.module) {
    const { module } = data;

    const idReplacer = replacer(() =>
      prepareId(
        module instanceof Module
          ? /** @type {ModuleId} */
            (/** @type {ChunkGraph} */ (chunkGraph).getModuleId(module))
          : module.id,
      ),
    );
    const moduleHashReplacer = hashLength(
      replacer(() =>
        module instanceof Module
          ? /** @type {ChunkGraph} */
            (chunkGraph).getRenderedModuleHash(module, data.runtime)
          : module.hash,
      ),
      // eslint-disable-next-line no-undefined
      'hashWithLength' in module ? module.hashWithLength : undefined,
      assetInfo,
      'modulehash',
    );
    const contentHashReplacer = hashLength(
      replacer(/** @type {string} */ (data.contentHash)),
      // eslint-disable-next-line no-undefined
      undefined,
      assetInfo,
      'contenthash',
    );

    replacements.set('id', idReplacer);
    replacements.set('modulehash', moduleHashReplacer);
    replacements.set('contenthash', contentHashReplacer);
    replacements.set(
      'hash',
      data.contentHash ? contentHashReplacer : moduleHashReplacer,
    );
    // Legacy
    replacements.set(
      'moduleid',
      deprecated(
        idReplacer,
        '[moduleid] is now [id]',
        'DEP_WEBPACK_TEMPLATE_PATH_PLUGIN_REPLACE_PATH_VARIABLES_MODULE_ID',
      ),
    );
  }

  // Other things
  if (data.url) {
    replacements.set('url', replacer(data.url));
  }
  if (typeof data.runtime === 'string') {
    replacements.set(
      'runtime',
      replacer(() => prepareId(/** @type {string} */ (data.runtime))),
    );
  } else {
    replacements.set('runtime', replacer('_'));
  }

  if (typeof path === 'function') {
    // eslint-disable-next-line no-param-reassign
    path = path(data, assetInfo);
  }

  // eslint-disable-next-line no-param-reassign
  path = path.replace(REGEXP, (match, content) => {
    if (content.length + 2 === match.length) {
      const contentMatch = /^(\w+)(?::(\w+))?$/.exec(content);
      if (!contentMatch) return match;
      const [, kind, arg] = contentMatch;
      // eslint-disable-next-line no-shadow
      const replacer = replacements.get(kind);
      // eslint-disable-next-line no-undefined
      if (replacer !== undefined) {
        return replacer(match, arg, path);
      }
    } else if (match.startsWith('[\\') && match.endsWith('\\]')) {
      return `[${match.slice(2, -2)}]`;
    }
    return match;
  });

  return path;
};

module.exports = replacePathVariables;
