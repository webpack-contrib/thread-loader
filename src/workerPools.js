import os from 'os';

import WorkerPool from './WorkerPool';

const workerPools = Object.create(null);

function calculateNumberOfWorkers() {
  // There are situations when this call will return undefined so
  // we are fallback here to 1.
  // More info on: https://github.com/nodejs/node/issues/19022
  const cpus = os.cpus() || { length: 1 };

  return Math.max(1, cpus.length - 1);
}

function getPool(options) {
  const workerPoolOptions = {
    name: options.name || '',
    numberOfWorkers: options.workers || calculateNumberOfWorkers(),
    workerNodeArgs: options.workerNodeArgs,
    workerParallelJobs: options.workerParallelJobs || 20,
    poolTimeout: options.poolTimeout || 500,
    poolParallelJobs: options.poolParallelJobs || 200,
    poolRespawn: options.poolRespawn || false,
  };
  const tpKey = JSON.stringify(workerPoolOptions);

  workerPools[tpKey] = workerPools[tpKey] || new WorkerPool(workerPoolOptions);

  return workerPools[tpKey];
}

export { getPool }; // eslint-disable-line import/prefer-default-export
