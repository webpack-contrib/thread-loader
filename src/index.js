import loaderUtils from 'loader-utils';

import { getPool } from './workerPools';

function pitch() {
  const options = loaderUtils.getOptions(this);
  const workerPool = getPool(options);

  if (!workerPool.isAbleToRun()) {
    return;
  }

  const callback = this.async();

  workerPool.run(
    {
      loaders: this.loaders.slice(this.loaderIndex + 1).map((l) => {
        return {
          loader: l.path,
          options: l.options,
          ident: l.ident,
        };
      }),
      resource: this.resourcePath + (this.resourceQuery || ''),
      sourceMap: this.sourceMap,
      emitError: this.emitError,
      emitWarning: this.emitWarning,
      loadModule: this.loadModule,
      resolve: this.resolve,
      getResolve: this.getResolve,
      target: this.target,
      minimize: this.minimize,
      resourceQuery: this.resourceQuery,
      optionsContext: this.rootContext || this.options.context,
      rootContext: this.rootContext,
    },
    (err, r) => {
      if (r) {
        r.fileDependencies.forEach((d) => this.addDependency(d));
        r.contextDependencies.forEach((d) => this.addContextDependency(d));
      }

      if (err) {
        callback(err);
        return;
      }

      callback(null, ...r.result);
    }
  );
}

function warmup(options, requires) {
  const workerPool = getPool(options);

  workerPool.warmup(requires);
}

export { pitch, warmup }; // eslint-disable-line import/prefer-default-export
