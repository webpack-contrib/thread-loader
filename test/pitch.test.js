import path from 'path';

import { pitch } from '../src/cjs';
import { getPool } from '../src/workerPools';

jest.mock('../src/workerPools', () => {
  return {
    getPool: jest.fn(),
  };
});

const runGetPoolMock = (error) => {
  getPool.mockImplementationOnce(() => {
    return {
      isAbleToRun: () => true,
      run: jest.fn((opts, cb) => {
        cb(error, {
          fileDependencies: [],
          contextDependencies: [],
          result: {},
        });
      }),
    };
  });
};

const runPitch = (options) =>
  pitch.call(
    Object.assign(
      {},
      {
        query: options,
        loaders: [],
        rootContext: path.resolve('../'),
        async: () => (error) => {
          if (error) {
            throw error;
          }
        },
      }
    )
  );

// it('runs pitch successfully when workPool not throw an error', () => {
//   runGetPoolMock(null);
//   expect(() => runPitch({})).not.toThrow();
// });

it('runs pitch unsuccessfully when workPool throw an error', () => {
  runGetPoolMock(new Error('Unexpected Error'));
  expect(() => runPitch({})).toThrowErrorMatchingSnapshot();
});
