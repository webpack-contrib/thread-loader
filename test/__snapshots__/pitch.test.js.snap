// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs pitch unsuccessfully when workPool throw an error 1`] = `"Unexpected Error"`;
